Add support for reflection of collation in types on PostgreSQL

dlax opened the pull request:
The 'collation' returned is `None` if it matches the default collation for the type. On the other hand, if the column collation matches the one of the database but is explicitly set at column creation, the value is reflected.

Related to #6511.

dlax committed:
Add support for reflection of collation in types on PostgreSQL

The 'collation' returned by PGDialect._columns_query() is None if it
matches the default collation for the type. On the other hand, if the
column collation matches the one of the database but is explicitly set
at column creation, the value is reflected.

Related to #6511.

CaselIT commented:
sorry for not yet getting a look at this. will try to come to it

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision baf48b55ee59874836c97607abfdfbf4f79fce4b of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
New Gerrit review created for change baf48b55ee59874836c97607abfdfbf4f79fce4b: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5993

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3801,6 +3802,19 @@ def _columns_query(self, schema, has_filter_names, scope, kind):
                     == pg_catalog.pg_attribute.c.attnum,
                 ),
             )
+            .join(
```
**Federico Caselli** (CaselIT) wrote:

Note to self: this could take advantage of the change in the style of the one added here https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5899

will try to check if it makes sense

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5993

dlax replied:
Added an extra commit to implement this idea (thus removing the `JOIN`s); and rebased.

dlax replied:
I think I fixed it; apparently `array_agg(<OIDs>)` requires a cast in pg8000.

No worries about other PRs, I'll rework them later.

CaselIT replied:
Yes, I had to do the same here https://github.com/sqlalchemy/sqlalchemy/commit/c3fb776b1f18e9a2a6058ddca20e03f16b513f44

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5993

dlax commented:
Fixed test failure for empty table case.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 62d03e71db4c459d49043a2e45e88f6489a02e77 of this pull request into gerrit so we can run tests and reviews and stuff

dlax commented:
Rebased to get a working flake8 setup (5c61cc6b93ab46f9a52408cba95b62a1fdcc9650).

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision c6ba75fe4ceb67a2636ffc2c432c7ce366ac0e9a of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset c6ba75fe4ceb67a2636ffc2c432c7ce366ac0e9a added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5993

dlax committed:
Cache pg_collation query to avoid joins

Following a similar approach to 703a323329b420fefec2b8a0a5f5f87ea3dc49d0.

The CAST(..., BIGINT) for array_agg()'s elements is needed for pg8000.

sqla-tester reviewed (commented):
**Denis Laxalde** (dlax) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5993

* lib/sqlalchemy/dialects/postgresql/base.py (line 3805): Done

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 4735f0ef65b4070abaab4f0632f691535803b7fd of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester reviewed (commented):
Hi, this is **sqla-tester** and I see you've pinged me for review. However, user **dlax** is not authorized to initiate CI jobs.  Please wait for a project member to do this!

dlax commented:
@CaselIT, would you mind pulling the latest version in gerrit to get CI jobs running? Thanks.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 896aa73356c7f5fb02d63f83c52ea1f45e7d3bd0 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 896aa73356c7f5fb02d63f83c52ea1f45e7d3bd0 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5993

CaselIT commented:
sure, sorry if this is taking a bit longer to merge.

dlax commented:
New version to fix an issue spotted by alembic tests.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 4c2ba7349824fdacc26c35d36f8b95cf3a35c99a of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 4c2ba7349824fdacc26c35d36f8b95cf3a35c99a added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5993

dlax committed:
Fix column collation reflection w.r.t. type's default in PostgreSQL

The previous `if` clause was wrong as it would match the `else` branch
if `default_collation_for_types` was `None` whereas we only want that
branch to match if the type is in `default_collation_for_types` array.

dlax committed:
Pass **kw to PGDialect._load_collation_dict()

Otherwise, reflection.cache() decorator does not work.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision c0a390314eb8ff2a4f729babf25cb052b6268d0a of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset c0a390314eb8ff2a4f729babf25cb052b6268d0a added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5993

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

gonna finish this one, thanks for the effort and sorry for the delay in getting to this

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5993

* lib/sqlalchemy/testing/requirements.py (line 695): should be closed

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

In the end using a case on a scalar subquery seem to have adequate performance without requiring a separate query

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5993

* lib/sqlalchemy/testing/requirements.py (line 695): Done

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

the new requirement was added to the publicly exported requirements so we should probably add a test to testing/suite/test_reflection.py also

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5993

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5993
